apollo-router/src/services/subgraph_service.rs

```diff
@@ -602,100 +604,205 @@ async fn call_http(
+                        .error(
+                            FetchError::SubrequestMalformedResponse {
+                                service: service_name.to_string(),
+                                reason: original_response,
+                            }
+                            .to_graphql_error(None),
+                        )
+                        .build()
+                })
+            })
+        }
+        (content_type, body, _) => {
+            // Something went wrong, compose a response with errors if they are present
+            let mut graphql_response = graphql::Response::builder().build();
+            if let Err(err) = content_type {
+                graphql_response.errors.push(err.to_graphql_error(None));
             }
+            if let Some(Err(err)) = body {
+                graphql_response.errors.push(err.to_graphql_error(None));
+            }
+            graphql_response
         }
+    };
 
-        let body = match hyper::body::to_bytes(body)
-            .instrument(tracing::debug_span!("aggregate_response_data"))
-            .await {
-                Err(err) => {
-
-                    tracing::error!(fetch_error = format!("{err:?}").as_str());
+    // Add an error for response codes that are not 2xx
+    if !parts.status.is_success() {
+        let status = parts.status;
+        graphql_response.errors.insert(
+            0,
+            FetchError::SubrequestHttpError {
+                service: service_name.to_string(),
+                status_code: Some(status.as_u16()),
+                reason: format!(
+                    "{}: {}",
+                    status.as_str(),
+                    status.canonical_reason().unwrap_or("Unknown")
+                ),
+            }
+            .to_graphql_error(None),
+        )
+    }
 
-                return Err(FetchError::SubrequestHttpError {
-                    status_code: None,
-                    service: service_name.clone(),
-                    reason: err.to_string(),
-                }.into())
+    let resp = http::Response::from_parts(parts, graphql_response);
+    Ok(SubgraphResponse::new_from_response(resp, context))
+}
 
-                }, Ok(body) => body,
-            };
+enum ContentType {
+    ApplicationJson,
+    ApplicationGraphqlResponseJson,
+}
 
+fn get_graphql_content_type(service_name: &str, parts: &Parts) -> Result<ContentType, FetchError> {
+    let content_type = parts
+        .headers
+        .get(header::CONTENT_TYPE)
+        .map(|v| v.to_str().map(MediaType::parse));
+    match content_type {
+        Some(Ok(Ok(content_type))) if (content_type.ty == APPLICATION && content_type.subty == JSON) => Ok(ContentType::ApplicationJson),
+        Some(Ok(Ok(content_type))) if (content_type.ty == APPLICATION && content_type.subty == GRAPHQL_RESPONSE && content_type.suffix == Some(JSON)) => Ok(ContentType::ApplicationGraphqlResponseJson),
+        Some(Ok(Ok(content_type))) => {
```

Comment:
nit: it may be more readable to match once on `Some(Ok(Ok(content_type)))` then have a nested match or branches on `content_type` fields

Reply:
This fixes cargo fmt so worth doing: [16de223](https://github.com/apollographql/router/pull/3328/commits/16de223f2ede8009d141c8d09114c3848682dba0)